architectures/M1/commands/upgrade.yml

```diff
@@ -5,82 +5,102 @@
 - import_playbook: "{{ tpa_dir }}/architectures/lib/init.yml"
   tags: always
 
-- name: Stop repmgr
+- name: Collect update facts for cluster {{ cluster_dir }}
   any_errors_fatal: true
   max_fail_percentage: 0
   become_user: root
   become: yes
   environment: "{{ target_environment }}"
-  hosts: all
+  hosts: "{{ update_hosts|default('all') }}"
   tasks:
-  - assert:
-      that:
-        failover_manager != 'patroni'
-      fail_msg:
-        Upgrade postgres does not support patroni managed clusters
-
-  - name: Stop repmgr
-    service:
-      name: repmgr
-      state: stopped
+  - include_role: name=postgres/facts
+    vars:
+      pgdata_initialised: true
     when: >
-      'postgres' in role
+      'postgres' in role and cluster_facts is not defined
 
-- name: Update postgres on replicas in cluster {{ cluster_dir }}
+- name: Update postgres on instances in M1 cluster {{ cluster_dir }}
```

Comment:
Why did the earlier code upgrade replicas first?

Reply:
IIUC the replicas are actually "upgraded" first, meaning only that the new binaries are installed with just a little bit of config + some failover config as well. When the primary upgrades last, it then upgrades the standbys in-place via `rsync`. 

The readme below outlines the M1 upgrade steps.
https://github.com/EnterpriseDB/solutions-architecture/blob/main/tpa/clusters/upgrade-postgres/cfg/M1/README.md